apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/ir/OperationBasedModelGroupBuilder.kt

```diff
@@ -174,16 +175,24 @@ internal class OperationBasedModelGroupBuilder(
       usedNames: MutableSet<String>,
       parentTypeConditions: List<String>,
   ): OperationField {
+    val rawTypeName = parentTypeConditions.lastOrNull() ?: ""
     if (selections.isEmpty()) {
       return OperationField(
           info = info,
           condition = condition,
           fieldSet = null,
+          shapes = emptyList(),
+          rawTypeName = rawTypeName
       )
     }
 
     val selfPath = path + "." + info.responseName
-
+    val shapes = shapes(
+        schema,
+        allFragmentDefinitions,
+        selections.map { it.selection }.filter { it !is GQLFragmentSpread },
```

Comment:
Can you elaborate on why you're filtering out fragment spreads? Feels like it'd be useful as well?  For an example, using the current code, 

```graphql
query TestQuery {
  hero {
    name
    ...DroidDetails
  }
}
```

Generates the below

```kotlin
   public data class Hero(
      public val __typename: String,
      public val name: String,
      public val droidDetails: DroidDetails?,
```

Feels like this is a place where `IHero` would have been handy as well?

Reply:
At first, I planed to include shapes from fragment spreads. 
```graphql
query TestOperation {
  something {
    ... ThingWithIDFragment
  }
}
fragment ThingWithIDFragment on ThingWithID {
  id
  ... on Human {
    name
  }
  ... on Droid {
    build_name
    model_name
  }
}
```
The operation above, there will be 2 shaped models which `thingWithIDFragment` field already shaped according possible type
```kotlin
public sealed interface ISomething {
  public val __typename: String,
  public val id: String,
  public val thingWithIDFragment: ThingWithIDFragment
}
// Shaped model: Human
public data class HumanSomething {
  public val __typename: String,
  public val id: String,
  public val thingWithIDFragment: HumanThingWithIDFragment // Shaped model of ThingWithIDFragment
}
// Shaped model: Droid
public data class DroidSomething {
  public val __typename: String,
  public val id: String,
  public val thingWithIDFragment: DroidThingWithIDFragment // Shaped model of ThingWithIDFragment
}
// Fallback model
public data class Something {
  public val __typename: String,
  public val id: String,
  public val thingWithIDFragment: ThingWithIDFragment
}
```
There are two reasons that made me decide not to do this.
After trying to implement, I found a problem with generated response adapter for `HumanSomething` and `DroidSomething`

Firstly, I couldn't make ImplementationAdapter work.

Reply:
Also, I don't normally write operation like this:
```graphql
query TestQuery {
  hero {
    name
    ...DroidDetails
  }
}
```
I would rather write it like below. I think it is not too difficult to get `IHero` if user wants to use it.
```graphql
query TestQuery {
  hero {
    name
    ... on Droid {
      ...DroidDetails
    }
  }
}
```

Reply:
> The operation above, there will be 2 shaped models which thingWithIDFragment field already shaped according possible type

I think "shapes()" is not what we want here. It's a bad name that caught me off guard [earlier](https://github.com/apollographql/apollo-kotlin/pull/4370#discussion_r963947262) actually. The `shapes()` function as in `shapes.kt` not only returns "user shapes" which might have no possible types but it also traverses the whole selection to find all possible type.

In your example: 

```graphql
query TestOperation {
  something {
    ... ThingWithIDFragment
  }
}
```

`something` is of type `ThingWithID` and only contains one fragment spread whose type condition is also `ThingWithID`. So the user should always see 

```kotlin
data class Something {
  public val __typename: String,
  public val id: String,
  public val thingWithIDFragment: ThingWithIDFragment # non-nullable because always satified
}
```

My earlier example is different though:

```graphql
query TestQuery {
  hero {
    name
    ...DroidDetails
  }
}
```

`DroidDetails` will not always be returned so we'd need 2 classes and 1 interface:

```kotlin
sealed interface IHero {
  val name: String
  val DroidDetails: DroidDetails? = null
}
class CharacterHero: IHero {
  val name: String
  val DroidDetails: DroidDetails? = null
}
class DroidHero: IHero {
  val name: String
  val DroidDetails: DroidDetails
}
```

So we need to enumerate all possible combinations of fragments. Which is different from the `shapes` from `shapes.kt` which is itself different from the Json shapes

Reply:
> I would rather write it like below.

That's more verbose. It's fine if people prefer this form but I wouldn't make that a recommendation.

Reply:
Anyway, it is just my opinion. If you think it's good to have, we can discuss more how to do it.

Reply:
> I think "shapes()" is not what we want here. It's a bad name that caught me off guard https://github.com/apollographql/apollo-kotlin/pull/4370#discussion_r963947262 actually. The shapes() function as in shapes.kt not only returns "user shapes" which might have no possible types but it also traverses the whole selection to find all possible type.

Do you mean we should not apply shape to fragment spread? but only apply non-null to fragment when its type condition always satisfied? 

Like this example...
```graphql
query TestOperation {
  something {
    ... ThingWithNameFragment
  }
}
fragment ThingWithNameFragment on ThingWithName {
  name
  ... on Human {
    friend {
      id
    }
  }
}
# schema
type Query {
  something: ThingWithID
}
interface ThingWithID {
  id: ID!
}

interface ThingWithName {
  name: String!
}
type Human implements ThingWithID & ThingWithName {
  id: ID!
  name: String!
  friend: ThingWithID!
}
type Pet implements ThingWithID & ThingWithName {
  id: ID!
  name: String!
}
type Droid implements ThingWithID {
  id: ID!
}
```
Operation above will generate code shown below...
``` kotlin
// shapes() output
// 0 = {Shape@2731} Shape(typeSet=[ThingWithID], possibleTypes=[Droid])
// 1 = {Shape@2732} Shape(typeSet=[ThingWithID, ThingWithName], possibleTypes=[Pet])
// 2 = {Shape@2733} Shape(typeSet=[ThingWithID, ThingWithName, Human], possibleTypes=[Human])

sealed interface ISomething {
  val __typename: String
  val thingWithNameFragment: ThingWithNameFragment?
}

// Shape 1
data class ThingWithNameSomething(
  val __typename: String,
  val thingWithNameFragment: ThingWithNameFragment
)
// Shape 2
data class HumanThingWithNameSomething(
  val __typename: String,
  val thingWithNameFragment: ThingWithNameFragment
)

// Fallback model
data class Something(
  val __typename: String,
  val thingWithNameFragment: ThingWithNameFragment?
)
```
Is this what you suggest?

Reply:
@martinbonnin I revised my code as my previous comment. Please check.

Reply:
@martinbonnin I have the feeling that there is no point to generate `Shape 2` in my [previous comment](https://github.com/apollographql/apollo-kotlin/pull/4370#discussion_r964356393) because it is identical to `Shape1` . 

I think I have to find a way to filter out shapes from inline fragment inside fragment spread.